sf_examples/megaverse_examples/megaverse/megaverse_utils.py

```diff
@@ -0,0 +1,109 @@
+import gym
+from megaverse.megaverse_env import MegaverseEnv, make_env_multitask
+
+from sample_factory.envs.env_utils import RewardShapingInterface, TrainingInfoInterface
+from sample_factory.utils.utils import log
+
+
+class MegaverseSpec:
+    def __init__(self, name):
+        self.name = name
+
+
+MEGAVERSE_ENVS = [
+    MegaverseSpec("megaverse_TowerBuilding"),
+]
+
+
+class Wrapper(gym.Wrapper, RewardShapingInterface, TrainingInfoInterface):
+    """Sets interface for PBT reward shaping, and also extra summaries for multi-task learning."""
+
+    def __init__(self, env, increase_team_spirit, max_team_spirit_steps):
+        gym.Wrapper.__init__(self, env)
+        RewardShapingInterface.__init__(self)
+        TrainingInfoInterface.__init__(self)
+
+        self.num_agents = env.unwrapped.num_agents
+        self.is_multiagent = env.unwrapped.is_multiagent
+
+        self.episode_rewards = [0] * self.num_agents
+
+        self.increase_team_spirit = increase_team_spirit
+        self.max_team_spirit_steps = max_team_spirit_steps
+
+        self.approx_total_training_steps = 0
+
+    def get_default_reward_shaping(self):
+        return self.env.unwrapped.get_default_reward_shaping()
+
+    def get_current_reward_shaping(self, agent_idx: int):
+        return self.env.unwrapped.get_current_reward_shaping(agent_idx)
+
+    def set_reward_shaping(self, reward_shaping: dict, agent_idx: int):
+        return self.env.unwrapped.set_reward_shaping(reward_shaping, agent_idx)
+
+    def reset(self, **kwargs):
+        self.episode_rewards = [0] * self.num_agents
+        return self.env.reset(), {}
+
+    def step(self, action):
+        obs, rewards, dones, infos = self.env.step(action)
+
+        for i, info in enumerate(infos):
+            self.episode_rewards[i] += rewards[i]
+
+            if dones[i]:
+                if "episode_extra_stats" not in info:
+                    info["episode_extra_stats"] = dict()
+                extra_stats = info["episode_extra_stats"]
+                extra_stats[f"z_{self.env.unwrapped.scenario_name.casefold()}_true_reward"] = info["true_reward"]
+                extra_stats[f"z_{self.env.unwrapped.scenario_name.casefold()}_reward"] = self.episode_rewards[i]
+
+                approx_total_training_steps = self.training_info.get("approx_total_training_steps", 0)
+                extra_stats["z_approx_total_training_steps"] = approx_total_training_steps
+
+                self.episode_rewards[i] = 0
+
+                if self.increase_team_spirit:
+                    rew_shaping = self.get_current_reward_shaping(i)
+                    rew_shaping["teamSpirit"] = min(approx_total_training_steps / self.max_team_spirit_steps, 1.0)
+                    self.set_reward_shaping(rew_shaping, i)
+                    extra_stats["teamSpirit"] = rew_shaping["teamSpirit"]
+
+        terminated = dones
+        truncated = dones
```

Comment:
This is not correct.
Megaverse currently does not differentiate between envs terminated by timer and by reaching terminal states.
So it is wrong to assume that all episodes finish by truncation (timeout). It should be `truncated=False`

It would be nice to get this information from the env, but definitely not right now :)